package NFeaturesRemap.NProto;

message TFeatutreDescr {
    required string Slice = 1;
    optional string Name = 2;
    optional uint32 Id = 3;
};

message TIf {
    required string Condition = 1;
    repeated TCalcDescr Set = 2;
};

message TCalcDescr {
    required string Expression = 1;
    required TFeatutreDescr FeatureDst = 2;
};

message TInitDescr {
    required string DefineVar = 1;
    required TFeatutreDescr FeatureSrc = 2;
};

enum EStage {
    // BEFORE_CLIENTS_ADJUST = 1;
    // BEFORE_L1 = 1;
    BEFORE_L2 = 2;
    BEFORE_NN_OVER_FEATURES = 3; //TODO: actual action is "before baesearch factors-stage response"
    // BEFORE_META_FEATURES = 4;
    AFTER_META_FEATURES_BEFORE_L3 = 5;
    // AFTER_L3_BEFORE_MIDDLE_REARRS = 6;
    // AFTER_MIDDLE_REARRS = 7;
    UNKNOWN_STAGE = 100;
};

message TCalculation {
    repeated string SearchTypes = 4;

    required EStage Stage = 1;
    repeated TInitDescr Let = 2;
    repeated TCalcDescr Set = 3;
    repeated TIf If = 5;
};

message TCalculations {
    repeated TCalculation StageCalculation = 1;
    optional bool AssumeThatAllVariablesAreConstantOverDocs = 2 [default = false];
}



